### ISSUE TITLE
Questions & Answers

### ISSUE THREAD

**Issue by ByBurton** [ByBurton/AI_Character_Editor#13]
Any questions about this editor go here please.

**Comment by PitchNeeded** [comment]
Would it be possible to have two stage siege attacks for AIs with AICE eg. a fast unit type like spearmen or macemen attacking with the archers and then the slow unit like or pikemen and swordsmen waiting and advancing a bit later.

**Comment by ByBurton** [comment]
Yes, if you set them as Additional Attack Unit 1 or 2. 
The main strike forces (primary & secondary) will always slow down to the slowest unit. So don't set spearmen and swordsmen there for example.

**Comment by J0J0Jo** [comment]
I might habe a workaround for the shields not being properly used. This working depends on the behaviour of the shields. Do they advance on their own? If they do that might work.

You could give an AI only shields as siege weapons and then choose e.g. archers as "AttUnitSiegeDef". Then they might stay close to the shields and might be protected while firing at the enemy castle. However, if the shields don't advance on their own this won't help as they would stay out of range.

Could this work? :)

**Comment by ByBurton** [comment]
Shields work in a weird manner. They go to a random friendly unit, preferring ranged units, no matter where they are. If they are back in base, well, guess where they are going...
Your suggestion should work.

**Comment by J0J0Jo** [comment]
Interesting.
So it kind of works for AIs with very few ranged defense and ranged patrol units (I am thinking of Philipp).

As the patch is out now I am going to try it out :)

**Comment by PitchNeeded** [comment]
Unfortunately it doesn't seem to be working for me at the moment at least. I tried changing phillip's spearmen to pikemen but he is still just recruiting spearmen :(

**Comment by ByBurton** [comment]
Did you put the created output.aic into the correct folder?

**Comment by PitchNeeded** [comment]
Yes I put it in its own aic folder in the Stronghold Crusader directory. 

Sultan is not sending slingers in his main attacks either as I tried to edit him to do :(

I am wondering if it is an administrator privileges issue for me....

**Comment by ByBurton** [comment]
What exception does the UCP throw, if any?

**Comment by PitchNeeded** [comment]
What do you mean by 'exception'?

**Comment by ByBurton** [comment]
What error message do you get if you install the UCP? And did you tick my aic file?

**Comment by Sh0wdown** [comment]
> Shields work in a weird manner. They go to a random friendly unit, preferring ranged units, no matter where they are. If they are back in base, well, guess where they are going...

That's interesting. Maybe I can fix that by adding a range limit to their target finding.

**Comment by PitchNeeded** [comment]
@ByBurton I don't get any error message at all and I did tick the button output.aic in the UCP

**Comment by ByBurton** [comment]
You should only tick one!!!!!! ONLY tick one at a time.

**Comment by PitchNeeded** [comment]
Ok I will try that.

**Comment by PitchNeeded** [comment]
It's working now, sorry for the inconvenience.

**Comment by Lolasik011** [comment]
@ByBurton And you can put the iron mines and quarries? I don't see this option here ...

**Comment by Rakso69** [comment]
What's the name of the Lord unit? Would like to try out an army of lords. Main reason behind that is because I want to keep the fear factor on an AI on the -5 for maximum efficency production, but also keeping the army strong. Another question tho: If one Lord "clone" die, does the AI die automatically? lmao
![Lords](https://user-images.githubusercontent.com/45049934/55718133-5210b200-59fb-11e9-8caf-cf370f1ce9ad.png)
What about any informations how to make all troops benefit from the Lords dots bonus instead? For example: Wolf AI Lord has 5 white dots. Could it be possible to make whole Wolf army use 5 white dots, so that units wouldn't be demoralized so much through the negative fear factor?

Back to my topic about the siege engines on UCP forum: Not all AIs start attacks from 1 tend, most of them do use 4 siege engines. I've got no clue how come my own edited AI makes only 1 damn tend... Tryed out everything, even turned back to the default settings of your Balance.aic file... Nothing... Such a shame. Frederick magically is used to besiege with 6 Trebuchets all the time. Mind exploded. lol

**Comment by ByBurton** [comment]
@Rakso69 This is an image I uploaded on discord, haha, glad you remembered it.
There is no name for the lord unit, and it is not supposed to be used.

**Comment by Rakso69** [comment]
@ByBurton lol, so how did you do the multiple lords then? :D Actually I've got the pic from @Heroesflorian - He's helping me with AIV.
What about the hidden third "fear factor" army strenght value lords use then? Could it be possible anyhow to add it on all units of the certain AI already with the AICE, or not yet?

Also, just out of curiosity, what does every single white dot do, and what's the overall +strenght effect of all 5 white dots together on Wolf Lord?

**Comment by Mizo5119** [comment]
Nice work. Enjoying the tools provided alot. I have a few questions because am a noob?

What exactly are sortie teams?
How can you increase iron mine and quarry count??

Thank you:)

**Comment by ByBurton** [comment]
@Mizo5119 Sortie units are units, that are sent out by the AI to protect farms. They are recruited with the highest priority of all units. Always.
You can't increase resource buildings yet.

@Rakso69 Ah, good old heroesflorian. We haven't found out what variable represents the strength of an AI lord. There are so many variables we don't know yet. Sorry, can't tell you much about this.

**Comment by ByBurton** [comment]
Also recruiting lords will not work. The AI won't use them.

**Comment by Sh0wdown** [comment]
> What about the hidden third "fear factor" army strenght value lords use then? Could it be possible anyhow to add it on all units of the certain AI already with the AICE, or not yet?

The lord strength is not adjustable through AIC edits.

**Comment by Rakso69** [comment]
More than 8 siege engines. Would it be possible soon?

**Comment by Grady99** [comment]
1) What does the RecruitInterval exactly do?
From vanilla settings(wolf ai):
RecruitInterval	= 4
RecruitIntervalWeak	= 1
RecruitIntervalStrong	= 4

From what I observed, the higher the number, the slower the recruitment. Rat for example has recruitment of 1 on strong and builds units a lot faster. Wouldn't this actually hinder the AI, for example when he establishes a good economy he's just gonna bank money instead of using it. 

2) Is it possible to set how many ox tethers the AI builds? Apart from wolf, every other AI builds too little and the stone just sits there. Meaning there's no use in letting him build more quarries if the currently built ones are not even cleared.

3) Is it possible to make the AI buy meat? (for the +1 popularity)

**Comment by DarkThief-Darek** [comment]
@Grady99 regarding your first question:

Your observations are correct. Some AIs get a handicap with the low recruitment speed, if they have a good economy. That's why the wolf for example was always so slow in attacking, because he slowly recruits units.

The funniest thing I observed was Nizar, who gets 8 on **RecruitIntervalWeak**. So he basically doesn't recruit units if nearly dead or at the start of the game.

**Comment by Mizo5119** [comment]
What does AttckRandom do? I get the fact that it randomly adds units into an attack, but like how does this work exactly in game xD?

Same thing goes for Raid.

**Comment by ByBurton** [comment]
@DarkThief-Darek Recruit interval being slow does not always have to be a bad thing. It is pretty vital for an AI to have money saved up. A broke AI, no matter how many units, is always gonna die at some point.
So it might not always be the wisest to spam units, except they are cheap enough.

@Mizo5119 AttackRandom works differently to the raid increase.
The first adds (additionally to the regular attack increase) a number of units from 0 to AttckRandom, randomized every time the attacker count is recalculated. At least that is what I remember it to be, could be slightly wrong. 
The raid increase works slightly different. Let's take The Rat for example. And let's say he has uses 13 raid units regularly, and has an increase of 10.
This value of 10 is the maximum random addition that is added towards the raiding units count.
There are special cases though: 
If the recruiting AI Lord has more than 5000 Gold, this value is doubled. If the AI Lord has less than 1000 Gold, it is set to 0. If the enemy of this lord (the one currently being attacked) has less than 500 Gold, this value is divided by MINUS 2, resulting in a decrease of raiding units.

Cases for the rat:
Rat has more than 5k Gold -> 13 + random(0 to 10) units
Rat has less than 1k Gold -> 13 + 0 units
Rat's enemy has less than 0.5k Gold -> 13 + random(0 to -5) units

**Comment by Mizo5119** [comment]
So basically it is a random increase of attack and raid team count that fluctuate based on specific conditions related to gold?

This can also explain the good reason to keep recruit speed for certain characters to be slow.
There would be a higher chance for attack and raid teams to be bigger.

I saw that you had The Pig have an attack team base count of 10 and a random increase up to a 100. Kinda curious with that decision since he is the only one with values like this.

Now just to keep it clear, the random team increase is separate from the regular attack base increase interval that can be set in the launcher right? So let's take Pig as an example. For his Attack if he 8has a base of 10 and a random increase of 100, with the attack increase interval of 0.5, then his next attack will be 15 +100 and so on? 

Another question regarding team compositions. I am still trying to figure out how Attackbackup and Attackranged and both the additional unit slots contribute to the total attack number. Are they seperate from the base number of team count ? If not, how are their numbers calculated? It's interesting as there is no base number for these teams, just a max number. Perhaps their base count is determined by the random attack increase number, meaning the higher it is, the more ranged a team will have.

Sorry for text wall.

**Comment by ByBurton** [comment]
@Mizo5119 

> So let's take Pig as an example. For his Attack if he 8has a base of 10 and a random increase of 100, with the attack increase interval of 0.5, then his next attack will be 15 +100 and so on?

Nope. It is random, reaching from 0 to 100. But The Pig is supposed to build massive armies, after attacking quickly the first time.

So if you HAVE enabled the percentage attack force increase in the patch installer, then this random increase will be ignored. For all AI lords.
If you have NOT enabled it, then the formula is BaseCount + (attack number + attack increase) + (0 to randomattackincrease). Attack increase being the constant value that you can choose instead of the percentage increase, it is 5 in vanilla.

Regarding your second question:
We are unsure of how the percentage is split, but the highest percentage is always the primary and secondary units.
RangedAttUnit s are units that actively try shooting down enemy defenders from walls, while BackupUnit s are units that stay back and protect the army while it is gathering, and only attack when the wall is breached.

**Comment by Grady99** [comment]
From my previous post:
> 1. Is it possible to set how many ox tethers the AI builds? Apart from wolf, every other AI builds too little and the stone just sits there. Meaning there's no use in letting him build more quarries if the currently built ones are not even cleared. (Edit: Idk I just saw that even wolf sometimes builds shit ton of them and sometimes too little like the other bots)

For some unknown reason if you make a custom castle for AI (with aiv editor) that makes the bot not build enough ox tethers, but not always! 

I am 100% sure as I spent an hour trying various combinations out to see if this really is the reason. For example rat builds 2 ox tethers per quarry, so if you have 5 quarry's he will be 10 tethers. But if you make a castle that meets condition x, that will cause **any** AI  (I didn't test for every single one but for several) to build very few of them in my cases for 6 quarries the AI always build like 3-5, and adding more quarries resulted in no additional tethers. But when I took an original wolf's castle and edited it a little(I kept most of it the same, maybe some optimisation here and there and some additional workshops) the AI wasn't affected at all and when I edited a little more he again built too little tethers. So some how the AIV affects the actual behaviour of the AI, at least in terms of the ox tether building.

I observed the same thing with Richard AI, no edits on his AIV or AIC at all, and sometimes he build the correct amount of tethers sometimes again only few. So again I'm not sure what is really causing this. But if you do a lot of editing in AIV the AI always then builds too little tethers, at least in my case.

If it will some how help the AI always adds tethers 1 by 1. Builds one, waits a little and keeps adding them slowly 1 by 1. In case of too little tethers he just builds few and stops there, never adding more.
